clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -319,23 +319,29 @@ public void disconnect(String nodeId) {
         if (connectionStates.isDisconnected(nodeId))
             return;
 
+        log.info("Client requested disconnect from node {} (invoking callbacks for in-flight requests)", nodeId);
```

Comment:
It seems we only have two types of callers for this: 1) admin-client proactively disconnecting, where the caller always log an info explaining why, 2) consumer-client async disconnect the coordinator.

If our target is primarily covering 2) here, what about: 1) make this function `disconnect(nodeId, reason)` where `reason` is a formatted string; 2) rename the `pendingDisconnects` (which is only used for coordinator disconnects) as `pendingCoordinatorDisconnects` and pass in a reason like "since the node was previously recognized as the group coordinator but now it become unknown".

Reply:
I think it is useful to cover both cases (as well as any future cases). Let me consider adding a disconnect reason.

Reply:
I found it a little simpler to add a separate log message for the coordinator case. Let me know if that seems reasonable. In general, I'm not too worried about the verbosity of this message since client-initiated disconnects should be relatively rare (e.g. after session or request timeouts).